validations/src/tests.rs

```diff
@@ -4844,64 +5106,168 @@ fn tally_dishonest_reward() {
+            expected: vec![],
+            found: vec![rewarded[2], rewarded[1], rewarded[0]],
+        }
+    )
+}
+
+#[test]
+fn tally_dishonest_reward() {
+    // Reveal value: integer(0)
+    let reveal_value = vec![0x00];
+    let liar_value = vec![0x0a];
+
+    // Create a DataRequestPool with 3 reveals (one of them is a lie from the data requester)
+    let dr_output = example_data_request_output_with_mode_filter(3, 200, 20);
+    let (dr_pool, dr_pointer, rewarded, slashed, error_witnesses, dr_pkh, _change, reward) =
+        dr_pool_with_dr_in_tally_stage_with_dr_liar(dr_output, 3, 3, 1, reveal_value, liar_value);
+    assert_eq!(reward, 200 + ONE_WIT + ONE_WIT / 2);
+
+    // Tally value: integer(0)
+    let tally_value = vec![0x00];
+    let vt0 = ValueTransferOutput {
+        time_lock: 0,
+        pkh: dr_pkh,
+        value: reward,
+    };
+    let vt1 = ValueTransferOutput {
+        time_lock: 0,
+        pkh: rewarded[1],
+        value: reward,
+    };
+    let vt2 = ValueTransferOutput {
+        time_lock: 0,
+        pkh: dr_pkh,
+        value: reward,
+    };
+    let tally_transaction = TallyTransaction::new(
+        dr_pointer,
+        tally_value,
+        vec![vt0, vt1, vt2],
+        slashed,
+        error_witnesses,
+    );
+    let x = validate_tally_transaction(&tally_transaction, &dr_pool, ONE_WIT, E).map(|_| ());
+
+    assert_eq!(
+        x.unwrap_err().downcast::<TransactionError>().unwrap(),
+        TransactionError::DishonestReward,
+    );
+}
+
+#[test]
+fn create_tally_validation_dr_liar() {
+    let reveal_value = RadonReport::from_result(
+        Ok(RadonTypes::from(RadonInteger::from(1))),
+        &ReportContext::default(),
+    );
+    let liar_value = RadonReport::from_result(
+        Ok(RadonTypes::from(RadonInteger::from(5))),
+        &ReportContext::default(),
+    );
+
+    // Create a DataRequestPool with 3 reveals (one of them is a lie from the data requester)
+    let dr_output = example_data_request_output_with_mode_filter(3, 200, 20);
+    let (dr_pool, dr_pointer, rewarded, _slashed, _error_witnesses, dr_pkh, _change, reward) =
+        dr_pool_with_dr_in_tally_stage_with_dr_liar(
+            dr_output.clone(),
+            3,
+            3,
+            1,
+            reveal_value.result.encode().unwrap(),
+            liar_value.result.encode().unwrap(),
+        );
+    assert_eq!(reward, 200 + ONE_WIT + ONE_WIT / 2);
+
+    // Create the RadonReport using the reveals and the RADTally script
+    let clause_result = evaluate_tally_precondition_clause(
+        vec![reveal_value.clone(), reveal_value, liar_value],
+        0.51,
+        3,
+        E,
+    );
+    let script = RADTally {
+        filters: vec![RADFilter {
+            op: RadonFilters::Mode as u32,
+            args: vec![],
+        }],
+        reducer: RadonReducers::Mode as u32,
+    };
+    let report = construct_report_from_clause_result(clause_result, &script, 3, E);
```

Comment:
If you are using an E > AFTER_SECOND_HARD_FORK, you should include the postcondition clause. (Same in other parts below)

Reply:
Done